### ISSUE TITLE
Sounds Do Not Play in Production

### ISSUE THREAD

**Issue by phindall** [jfversluis/Plugin.Maui.Audio#142]
OK.  Poking my eyes out on this one.  I have a Maui iOS app where I want to play some simple sounds for user feedback (-e.g. ding).  I had no trouble adding Plugin.Maui.Audio and using it to produce sounds.  Sounds are audible on iOS simulator, and on physical iPhones connected to development Mac during debuging.  But, when the app is built in release and published to the App Store, then downloaded to a phone, the sounds no longer play.  I know your (and my) first guess is that there must be something wrong with the phone where the published app is installed (volume is turned down, speaker is broken, etc.) but we have tested multiple phones and all exhibit the same behavior.  To help troubleshoot, I created a simple Maui page with a single button and event handler like this.

```
private async void PlayDing_Clicked(object sender, EventArgs e)
{
   IAudioPlayer player = AudioManager.Current.CreatePlayer(await FileSystem.OpenAppPackageFileAsync("ding.wav"));
   player.Play();
}
```

Again, the above works fine on an iOS simulator, and on a physical phone during debug, but not in a released version of the app.

Although not shown above, I did verify that the the ding.wav file is found and a readable stream is created from FileSystem.OpenAppPackageFileAsync("ding.wav").  So, I am confident the file is part of the released app package.  Could it be that some phones cannot play wav files?  Are there any missing entitlements required to play audio?  I'm really sorry if I've done something stupid, but I'm at at loss.

**Comment by bijington** [comment]
I suspect it is because your `player` variable is going out of scope before the sound gets played. Can you try changing your code to

```csharp

private async void PlayDing_Clicked(object sender, EventArgs e)
{
   IAudioPlayer player = AudioManager.Current.CreateAsyncPlayer(await FileSystem.OpenAppPackageFileAsync("ding.wav"));
   await player.Play();
}

```

**Comment by phindall** [comment]
Thanks for the quick reply.  I did try changing the code and it did not make a difference.  I note that you have a couple errors in it, but I get the gist.  The correct code is:
```
private async void PlayDing_Clicked(object sender, EventArgs e)
{
    AsyncAudioPlayer player = AudioManager.Current.CreateAsyncPlayer(await FileSystem.OpenAppPackageFileAsync("ding.wav"));
    await player.PlayAsync(new CancellationToken(false));
}
```
Unfortunately, this did not fix the problem.  But I think I figured it out.  And, if I can make a plea to myself and other developers who like to stay up late, drink beer and hack away at problems until the wee hours. . . stop and get some sleep.  Things will be clearer in the morning!  

I believe the underlying problem has something to do with differences in v3.0.0 and v3.0.1.  I don’t know what, but at least I now have consistent behavior between debug and app-store released versions.  I feel so much less insane 😊.

Some explanation here is required.  Our MAUI app is a white label template that we re-skin and customize for various clients.  To accomplish this, we develop & debug in a single VS solution, then merge code files into other VS solutions which contain the client’s customizations (colors, fonts, iconography, etc.).  These secondary solutions are ultimately the basis for release as independent app packages into the app stores.  The problem was that somehow (most likely one of my team members) upgraded the Plugin.Maui.Audio NuGet package on the client’s solution to v3.0.1.  However, our development & debug solution was still assigned to v3.0.0.  This was the reason for the difference in behavior between debug and production.

I downgraded the client’s solution to v3.0.0, republished to the AppStore and my sounds returned!  For fun, I also tried upgrading our development project to v3.0.1 and my sounds disappeared in debugging.  This is true whether you try your code example, or my original code.  Is there something amiss with v3.0.1?  We will stay with 3.0.0 for now which appears to work as we expect.

**Comment by bijington** [comment]
Thanks for the detailed response and fixing the shoddy code I wrote in my phone 😀 

There really aren't many changes between 3.0.0 and 3.0.1 but the most likely change is the bump from `net8.0-ios17.2` to `net8.0-ios17.5`. This isn't something that we explicitly changed but comes from the wonders of workloads. Can you confirm what MAUI workloads you have installed on the release/build machines?

**Comment by phindall** [comment]
You know, open source and Agile development has made development far too complex, for no reason.  There are simply too many interdependent components (Mono, .NET, Maui, 3rd party NuGet packages, plus Android and Apple), all on their own separate never-ending upgrade path.  I can't shake the feeling that all my MAUI development is built on a house of cards that could crash at any moment.

I've been writing software for 40 years and I don't even know what a "workload" is.  Like everything, I had to Google it to figure out how to get you the answer.  Yes, apparently the workload on my build machine is 17.2.  One day, someone is going to have to explain to me how all this stuff works.

Man, I miss building WinForms applications. 😂.

**Comment by bijington** [comment]
I don't agree that open-source and agile development are the causes of the complexities. The nature of cross-platform development is not a simple problem to solve. That said workloads don't seem to help.

Sadly workloads are the cause of some pain, in an ideal world we could remove this dependency layer but there does need to be SDK support overlapping with nugets downloaded. As library authors we have discovered that we can explicitly target TFMs like `net8.0-ios17.2` instead of `net8.0-ios` and then rely on the magic of workloads installed to determine the specific TFM being targeted.

To give some further explanation of the issue the following may help.

If we inspect the `3.0.0` [Nuget package](https://nuget.info/packages/Plugin.Maui.Audio/3.0.0) you can see
<img width="1487" alt="Screenshot 2024-11-26 at 19 39 39" src="https://github.com/user-attachments/assets/9ac5e704-c1b0-44e2-98eb-4c8cef591064">

If we inspect the `3.0.1` [Nuget package](https://nuget.info/packages/Plugin.Maui.Audio/3.0.1) you can see
<img width="1486" alt="Screenshot 2024-11-26 at 19 39 53" src="https://github.com/user-attachments/assets/2446cb57-e1ca-43d1-9489-3480224ecdc0">

This is a result of us implicitly targeting a TFM rather than an explicit version. I haven't seen any official documentation on this topic to help library authors though.

**Comment by billylo1** [comment]
> AsyncAudioPlayer player = AudioManager.Current.CreateAsy

Side note of other readers:  My synchronous AudioPlayer has recently stopped playing on both iOS and Android. I was not able to track down the root cause (probably with .NET 8 / MAUI workload updates.) But your suggestion to go with Async made everything work again.  Thanks!

**Comment by jfversluis** [comment]
Might be related to #136 and some trimming/aot stuff?

**Comment by smstosandeep** [comment]
There is indeed something. I upgraded the solution(.net 9) some time back. Recently, I started testing recording and playing options of my app.  Recording was working but playing stopped working in release mode. In debug everything is still working. Tried a lot of things async/sync, rewrite code but did not work.

I thought what have i done !!
Then i downgraded to version 3.0.0 and it started  working. Just tested while writing. This is the piece of code i am/was having in the app.
          `       var stream = KFileHelper.GetFileStream(filePath);

                if (stream != null)
                {
                    player = AudioManager.Current.CreateAsyncPlayer(stream);
                    await player.PlayAsync(CancellationToken.None);
                }
`
